### PR TITLE
Add in user route functionality

### PR DESCRIPTION
**Pull request by ABarroso647**
Add in all functionality to create, update, and delete users (on their own)

### CODE REVIEW

File: backend/controller/user.go [hunk 1]
```diff
@@ -2,14 +2,87 @@ package controller
...
+	if result.Error != nil {
+		helpers.DBErrorHandling(result.Error, w, r)
+		return
+	}
+	//go yells at you if you don't error check this, better to be safe
+	errJson := json.NewEncoder(w).Encode(&users)
+	if errJson != nil {
+		http.Error(w, "500", http.StatusInternalServerError)
+		return
+	}
```
Comment (AdrianJendo) [inline backend/controller/user.go:26]: how is data returned?

File: backend/controller/user.go [hunk 2]
```diff
@@ -2,14 +2,87 @@ package controller
...
+	errJson := json.NewEncoder(w).Encode(&user)
+	if errJson != nil {
+		http.Error(w, "500", http.StatusInternalServerError)
+		return
+	}
+}
+
+func CreateUser(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	var user models.UserReq
```
Comment (AdrianJendo) [inline backend/controller/user.go:47]: What is UserReq vs. User?

File: backend/controller/user.go [hunk 3]
```diff
@@ -2,14 +2,87 @@ package controller
...
+		if err != nil {
+			log.Print(err)
+		} else {
+			log.Print("Request doesn't contain all params")
+		}
+		http.Error(w, "400", http.StatusBadRequest)
+		return
+	}
+	dbUser := models.User{Name: user.Name,
+		Email: user.Email, IsAdmin: user.IsAdmin}
```
Comment (AdrianJendo) [inline backend/controller/user.go:59]: In the future can we do this format:
```
models.User {
    Name: user.Name,
    Email: user.Email,
    IsAdmin: user.IsAdmin,
}
```

File: backend/controller/user.go [hunk 4]
```diff
@@ -2,14 +2,87 @@ package controller
...
+	result := config.DB.Create(&dbUser)
+	if result.Error != nil {
+		helpers.DBErrorHandling(result.Error, w, r)
+		return
+	}
+	errJson := json.NewEncoder(w).Encode(r.Host + "/api/users/{" + strconv.Itoa(int(dbUser.ID)) + "}")
```
Comment (AdrianJendo) [inline backend/controller/user.go]: What does this do?

File: backend/models/user.go [hunk 5]
```diff
@@ -4,8 +4,13 @@ import "gorm.io/gorm"
...
-	IsAdmin bool   `json:"is_admin"`
+	Name    string
+	Email   string
+	IsAdmin bool `json:"default;False"`
+}
+
+type UserReq struct {
+	Name    string `json:"name,omitempty"`
+	Email   string `json:"email,omitempty"`
+	IsAdmin bool   `json:"is_admin,string,default;False"`
```
Comment (AdrianJendo) [inline backend/models/user.go:15]: Is it not possible to combine these into one `User` struct? Why do we have 2.

File: backend/controller/user.go (hunk 2 above)
Comment (ABarroso647) [reply backend/controller/user.go:47]: Its in models, essentially userReq contains a list of all the parameters that can be inputted to, (name, email, IsAdmin) and not the non-inputtable params, this way during user input we can verify against the proper params

File: backend/models/user.go (hunk 5 above)
Comment (ABarroso647) [reply backend/models/user.go:15]: See other comment

File: backend/controller/user.go (hunk 4 above)
Comment (ABarroso647) [reply backend/controller/user.go]: returns a get request link with the newly created userId that can be hit to get the user object, can add a comment for it

File: backend/controller/user.go (hunk 1 above)
Comment (ABarroso647) [reply backend/controller/user.go:26]: its not in these lines, but line 21, when we try to write &users, that is what is returning the data

File: backend/controller/user.go (hunk 3 above)
Comment (ABarroso647) [reply backend/controller/user.go:59]: Yes

File: backend/controller/user.go (hunk 1 above)
Comment (AdrianJendo) [reply backend/controller/user.go:26]: line 23

File: backend/controller/user.go (hunk 4 above)
Comment (AdrianJendo) [reply backend/controller/user.go]: So we are executing a GET request to get the user object when returning data to the client? We can't just grab the user object directly here?

File: backend/controller/user.go (hunk 2 above)
Comment (AdrianJendo) [reply backend/controller/user.go:47]: fine

File: backend/controller/user.go (hunk 4 above)
Comment (ABarroso647) [reply backend/controller/user.go]: we could if thats the wanted functionality

Comment (ABarroso647) [reply backend/controller/user.go]: depends on the intended use case, I can update it as you think you need

Comment (AdrianJendo) [reply backend/controller/user.go]: This just seems unnecessary steps to me, ill let cameron decide what to do

Comment (ABarroso647) [reply backend/controller/user.go]: he said he'll look into it

Comment (cameronkinsella) [reply backend/controller/user.go]: Apparently the best practice for this is to return the newly created/updated object in the same format you would for a GET request for that object.

Comment (AdrianJendo) [reply backend/controller/user.go]: But can we achieve this without kicking off another request? I've never seen this pattern of actually calling a GET? Like wouldn't even just calling `GetAllUsers` be better?

Comment (cameronkinsella) [reply backend/controller/user.go]: To clarify, I am not talking about kicking off another request, I'm talking about avoiding one.
For example:
POST `/user` creates user with ID 123. Creating this user (the response of the POST request) would return the same response as GET `/user/123`, so you would not need to call  GET `/user/123` after creating it to get its information.

**Review by AdrianJendo — APPROVED** [review]
That's much better lol

**Pull request merged**